### PR TITLE
Trafodion 2001

### PR DESCRIPTION
**Pull request by zcorrea**
This is an initial pull request to validate changes which enable elasticity changes (JIRA: TRAFODION-2001).
[https://issues.apache.org/jira/browse/TRAFODION-2001](url)
These elasticity changes allow for Trafodion cluster node (hostname) membership to be changed while the instance is operational. Nodes can be added or deleted while the instance is running. 

These changes include:

o Implementation of Trafodion Configuration API shared library
  - Provides an interface to a repository for Trafodion Configuration and Registry
     + Initial repository storage method implemented is SQLite
  - Allows for future repository storage methods such as Zookeeper, MySQL, etc. to added without affecting the rest of the Trafodion components
o Implementation of 'PERSIST' configuration section
  - A template based persistent process definition
o Implementation of new shell commands (and enhancements to existing commands):
  - node add node-name \<node-name\>,
              cores {\<first-core\>} [ - \<last-core\>}],
              processors {\<processor-count\>},
              roles {connection|aggregation|storage}
 - node config [\<nid\>|\<node-name\>]
 - node delete \<node-name\>
 - node down \<nid\>|\<node-name\> [, \<reason-string\>]
 - node info [\<nid\>]
 - node name \<old-node-name\> \<new-node-name\>
 - node up \<name\>
 - persist config [{keys}|\<persist-process-prefix\>]
 - persist exec \<persist-process-prefix\>
 - persist info [\<persist-process-prefix\>]
 - persist kill \<persist-process-prefix\>

### CODE REVIEW

**Comment by svarnau** [comment]
Wow, over 200 files changed.  I don't suppose there is any way to break this into smaller chunks?
Maybe we should discuss on dev list how folks want to review this.

**Comment by svarnau** [comment]
I see changes to bash installer, which is deprecated already. python-installer is being used for testing.

File: core/sqf/sql/scripts/sqnodestatus [hunk 1]
```diff
@@ -26,12 +26,11 @@
...
 #   ssh -M "ConnectTimeout 1" -o "ConnectionAttempts 3" n028 hostinfo 2>&1
 #
-# Uses:
-#   export MY_NODES=' -w n028 -w n044 -w n029 -w n045 -w n030 -w n031 -w n032 '
+# Uses the following enviroment variables:
 #   export SQ_MON_SSH_OPTIONS=' -o "ConnectTimeout 1" -o "ConnectionAttempts 3" '
 #
 
-my $node_context=lc(readpipe("echo \$NODE_LIST"));
+my $node_context=readpipe("trafconf -name");
```
Comment (mkby) [inline core/sqf/sql/scripts/sqnodestatus]: will `trafconf -name` return the lowercase hostname or the upper/lower case of hostname is addressed in other codes?

Comment (zcorrea) [inline core/sqf/sql/scripts/sqnodestatus]: The upper lower case is addressed in other code. Specifically, in the monitor code. The trafconf utility is new and was created to address the need of obtaining the current hostname (node name) membership in a Trafodion cluster instance. I will add more details to its purpose and usage in the JIRA: TRAFODION-2001. A brief summary is as follows:

- Node membership is initially derived from 'sqconfig' and used to populate the Trafodion Configuration database. After this, node membership is through the 'shell' node add/delete commands.
- The Trafodion Configuration database is the configuration of record once the instance is started.
- The 'trafconf' utility reads the contents of the Trafodion Configuration database to display the current node membership of the cluster.
- The 'trafconf' utility us used by scripts and by the Trafodion user .bashrc to populate environment variables with the current node membership. Specifically, the MY_NODES environment variable.

Comment (mkby) [inline core/sqf/sql/scripts/sqnodestatus]: looks good!

File: install/python-installer/scripts/traf_sqconfig.py [hunk 2]
```diff
@@ -51,18 +52,15 @@ def run():
...
-
-    lines.append('end overflow\n')
 
+    # write out the node section
     with open(sqconfig_file, 'w') as f:
         f.writelines(lines)
 
+    # append the current persist section to the Trafodion configuration file
+    run_cmd('echo "sed \'1,20d\' %s >> %s"' % (sqconfig_persist_file, sqconfig_file))
+    run_cmd('sed \'1,20d\' %s >> %s' % (sqconfig_persist_file, sqconfig_file))
```
Comment (svarnau) [inline install/python-installer/scripts/traf_sqconfig.py]: why are we stripping header comments out?
seems like we ought to include file verbatim.

Comment (svarnau) [inline install/python-installer/scripts/traf_sqconfig.py]: In fact, why are the installers concatenating these files?
Why doesn't monitor just directly read both sqconfig and sqconfig.persist?
Or sqgen take some file name parameters?

Comment (zcorrea) [inline install/python-installer/scripts/traf_sqconfig.py]: On stripping the header comments out, the target file already has the same Copyright information.

On concatenating, yes, the sqgen should probably be modified to pull in the persist file directly after processing the node section on the 'sqconfig' file. Good suggestion!

**Comment by zcorrea** [comment]
Regarding the Jenkins test failure, I am seeking SQL developer help to correct the failure of core-regress-core-hdp.

Perhaps someone has seen this failure and would know how to correct the problem?

The following is the diff file contents:

cat DIFF116
946c946,958
< --- 1 row(s) inserted.
---
> *** ERROR[8448] Unable to access Hbase interface. Call to ExpHbaseInterface::checkAndInsertRow returned error HBASE_ACCESS_ERROR(-706). Cause: java.io.IOException: getTmId: IdTm threw exception
> org.apache.hadoop.hbase.client.transactional.RMInterface.getTmId(RMInterface.java:240)
> org.apache.hadoop.hbase.client.transactional.RMInterface.checkAndPutRegionTx(RMInterface.java:402)
> org.TRAFODION.sql.HTableClient.putRow(HTableClient.java:1585)
> org.TRAFODION.sql.HBaseClient.insertRow(HBaseClient.java:1729) Caused by
> org.apache.hadoop.hbase.regionserver.transactional.IdTmException: id threw:org.apache.hadoop.hbase.regionserver.transactional.IdTmException: ferr=11
> org.apache.hadoop.hbase.regionserver.transactional.IdTm.id(IdTm.java:134)
> org.apache.hadoop.hbase.client.transactional.RMInterface.getTmId(RMInterface.java:236)
> org.apache.hadoop.hbase.client.transactional.RMInterface.checkAndPutRegionTx(RMInterface.java:402)
> org.TRAFODION.sql.HTableClient.putRow(HTableClient.java:1585)
> org.TRAFODION.sql.HBaseClient.insertRow(HBaseClient.java:1729).
>
> --- 0 row(s) inserted.
956,958c968
< *** ERROR[8616] A conflict was detected during commit processing. Transaction has been aborted. Detail @conflict details@
<
< --- SQL operation failed with errors.
---
> --- SQL operation complete.
964c974
<  1 2
---
>  1 1
1054c1064,1076
< --- 1 row(s) inserted.
---
> *** ERROR[8448] Unable to access Hbase interface. Call to ExpHbaseInterface::insertRow returned error HBASE_ACCESS_ERROR(-706). Cause: java.io.IOException: getTmId: IdTm threw exception
> org.apache.hadoop.hbase.client.transactional.RMInterface.getTmId(RMInterface.java:240)
> org.apache.hadoop.hbase.client.transactional.RMInterface.putRegionTx(RMInterface.java:365)
> org.TRAFODION.sql.HTableClient.putRow(HTableClient.java:1599)
> org.TRAFODION.sql.HBaseClient.insertRow(HBaseClient.java:1729) Caused by
> org.apache.hadoop.hbase.regionserver.transactional.IdTmException: id threw:org.apache.hadoop.hbase.regionserver.transactional.IdTmException: ferr=22
> org.apache.hadoop.hbase.regionserver.transactional.IdTm.id(IdTm.java:134)
> org.apache.hadoop.hbase.client.transactional.RMInterface.getTmId(RMInterface.java:236)
> org.apache.hadoop.hbase.client.transactional.RMInterface.putRegionTx(RMInterface.java:365)
> org.TRAFODION.sql.HTableClient.putRow(HTableClient.java:1599)
> org.TRAFODION.sql.HBaseClient.insertRow(HBaseClient.java:1729).
>
> --- 0 row(s) inserted.
1064,1066c1086
< *** ERROR[8616] A conflict was detected during commit processing. Transaction has been aborted. Detail @conflict details@
<
< --- SQL operation failed with errors.
---
> --- SQL operation complete.
1072c1092
<  1 2
---
>  1 1
1163c1183,1195
< --- 1 row(s) deleted.
---
> *** ERROR[8448] Unable to access Hbase interface. Call to ExpHbaseInterface::deleteRow returned error HBASE_ACCESS_ERROR(-706). Cause: java.io.IOException: getTmId: IdTm threw exception
> org.apache.hadoop.hbase.client.transactional.RMInterface.getTmId(RMInterface.java:240)
> org.apache.hadoop.hbase.client.transactional.RMInterface.deleteRegionTx(RMInterface.java:338)
> org.TRAFODION.sql.HTableClient.deleteRow(HTableClient.java:1391)
> org.TRAFODION.sql.HBaseClient.deleteRow(HBaseClient.java:1776) Caused by
> org.apache.hadoop.hbase.regionserver.transactional.IdTmException: id threw:org.apache.hadoop.hbase.regionserver.transactional.IdTmException: ferr=11
> org.apache.hadoop.hbase.regionserver.transactional.IdTm.id(IdTm.java:134)
> org.apache.hadoop.hbase.client.transactional.RMInterface.getTmId(RMInterface.java:236)
> org.apache.hadoop.hbase.client.transactional.RMInterface.deleteRegionTx(RMInterface.java:338)
> org.TRAFODION.sql.HTableClient.deleteRow(HTableClient.java:1391)
> org.TRAFODION.sql.HBaseClient.deleteRow(HBaseClient.java:1776).
>
> --- 0 row(s) deleted.
1173,1175c1205
< *** ERROR[8616] A conflict was detected during commit processing. Transaction has been aborted. Detail @conflict details@
<
< --- SQL operation failed with errors.
---
> --- SQL operation complete.
1267c1297,1309
< --- 1 row(s) deleted.
---
> *** ERROR[8448] Unable to access Hbase interface. Call to ExpHbaseInterface::deleteRow returned error HBASE_ACCESS_ERROR(-706). Cause: java.io.IOException: getTmId: IdTm threw exception
> org.apache.hadoop.hbase.client.transactional.RMInterface.getTmId(RMInterface.java:240)
> org.apache.hadoop.hbase.client.transactional.RMInterface.deleteRegionTx(RMInterface.java:338)
> org.TRAFODION.sql.HTableClient.deleteRow(HTableClient.java:1391)
> org.TRAFODION.sql.HBaseClient.deleteRow(HBaseClient.java:1776) Caused by
> org.apache.hadoop.hbase.regionserver.transactional.IdTmException: id threw:org.apache.hadoop.hbase.regionserver.transactional.IdTmException: ferr=11
> org.apache.hadoop.hbase.regionserver.transactional.IdTm.id(IdTm.java:134)
> org.apache.hadoop.hbase.client.transactional.RMInterface.getTmId(RMInterface.java:236)
> org.apache.hadoop.hbase.client.transactional.RMInterface.deleteRegionTx(RMInterface.java:338)
> org.TRAFODION.sql.HTableClient.deleteRow(HTableClient.java:1391)
> org.TRAFODION.sql.HBaseClient.deleteRow(HBaseClient.java:1776).
>
> --- 0 row(s) deleted.
1277,1279c1319
< *** ERROR[8616] A conflict was detected during commit processing. Transaction has been aborted. Detail @conflict details@
<
< --- SQL operation failed with errors.
---
> --- SQL operation complete.
1376c1416,1428
< --- 1 row(s) updated.
---
> *** ERROR[8448] Unable to access Hbase interface. Call to ExpHbaseInterface::insertRow returned error HBASE_ACCESS_ERROR(-706). Cause: java.io.IOException: getTmId: IdTm threw exception
> org.apache.hadoop.hbase.client.transactional.RMInterface.getTmId(RMInterface.java:240)
> org.apache.hadoop.hbase.client.transactional.RMInterface.putRegionTx(RMInterface.java:365)
> org.TRAFODION.sql.HTableClient.putRow(HTableClient.java:1599)
> org.TRAFODION.sql.HBaseClient.insertRow(HBaseClient.java:1729) Caused by
> org.apache.hadoop.hbase.regionserver.transactional.IdTmException: id threw:org.apache.hadoop.hbase.regionserver.transactional.IdTmException: ferr=11
> org.apache.hadoop.hbase.regionserver.transactional.IdTm.id(IdTm.java:134)
> org.apache.hadoop.hbase.client.transactional.RMInterface.getTmId(RMInterface.java:236)
> org.apache.hadoop.hbase.client.transactional.RMInterface.putRegionTx(RMInterface.java:365)
> org.TRAFODION.sql.HTableClient.putRow(HTableClient.java:1599)
> org.TRAFODION.sql.HBaseClient.insertRow(HBaseClient.java:1729).
>
> --- 0 row(s) updated.
1386,1388c1438
< *** ERROR[8616] A conflict was detected during commit processing. Transaction has been aborted. Detail @conflict details@
<
< --- SQL operation failed with errors.
---
> --- SQL operation complete.
1394c1444
<  1 111
---
>  1 333

**Comment by carolp222** [comment]
I've been running some experiments with this code, looking at availability issues. I've been downing the node via sqshell, rebooting, etc..  Mostly it seems to be pretty solid - thanks!  One issue keeps coming up though:  When I up a node in sqshell after a failure, I consistently see that RMS processes don't get restarted.  Then, because the RMS processes don't get started, connectivity won't restart the missing mxosrvrs

**Comment by zcorrea** [comment]
jenkins, restest

**Comment by zcorrea** [comment]
jenkins, retest

**Comment by zcorrea** [comment]
jenkins, retest

**Comment by zcorrea** [comment]
jenkins, retest

**Comment by zcorrea** [comment]
jenkins, retest

**Comment by zcorrea** [comment]
jenkins, retest

**Comment by carolp222** [comment]
I've run this code on a couple of platforms, rebooting servers and upping and downing nodes while running transactional and non-transactional workloads.  I've pushed it a bit and verified that my workloads completed successfully.  At this point, it seems pretty solid.

**Comment by sureshsubbiah** [comment]
+1. I have not reviewed the change. The vote is based on Eason, Steve and Carol's reviews. Merging

**Pull request merged**